File: for-jython/udrawingsurface.py
# unit Udrawingsurface

import u3dsupport
import delphi_compatability
from conversion_common import *

#import utdo
#import ucollect

# const
kInitialDrawingSurfaceTriangles = 1000


# ----------------------------------------------------------------------------- *KfDrawingSurface creating/destroying 

class KfDrawingSurface:
    def __init__(self):
        self.drawingContext = None
        self.lineContrastIndex = 0
        self.drawingLines = False
        self.circlingPoints = False
        self.foreColor = delphi_compatability.EmptyColor
        self.backColor = delphi_compatability.EmptyColor
        self.lineColor = delphi_compatability.EmptyColor
        self.circleColor = delphi_compatability.EmptyColor
        self.triangles = []
        self.recording = False
        self.lineWidth = 0.0
        self.circlePointRadius = 0
        self.sortTdosAsOneItem = False
    
        self.triangles = []
        self.fillingTriangles = True
        self.initialize()
        # start with a bunch of triangles 
        for i in range(0, kInitialDrawingSurfaceTriangles + 1):
            self.triangles.append(u3dsupport.KfTriangle())
    
    def setDrawingContext(self, drawingContext):
        self.drawingContext = drawingContext

    def initialize(self):
        #backColor := clLtGray;
        #  foreColor := clLtGray;
        #  lineColor := clLtGray;
        self.lineWidth = 1.0
        self.recording = False
        self.clearTriangles()
        #if self.drawingContext:
        #    if self.fillingTriangles:
        #        self.drawingContextUseSolidBrush()
        #    else:
        #        self.drawingContextUseClearBrush()
    
    # ----------------------------------------------------------------------------- KfDrawingSurface managing triangles 
    def clearTriangles(self):
        self.triangles = []
    
    def plantPartIDForPoint(self, point):
        points = []
        for i in range(2):
            points.append(u3dsupport.KfPoint3D())

        x = point.x
        y = point.y
        result = -1
        closestDistanceSoFar = 0
        closestPartID = -1
        for triangle in self.triangles:
            if triangle.isLine:
                centerX = (triangle.points[0].x + triangle.points[1].x) / 2.0
                centerY = (triangle.points[0].y + triangle.points[1].y) / 2.0
                thisDistance = (centerX - x) * (centerX - x) + (centerY - y) * (centerY - y)
                if (closestPartID == -1) or (thisDistance < closestDistanceSoFar):
                    closestDistanceSoFar = thisDistance
                    closestPartID = triangle.plantPartID
            else:
                points[0].X = intround(triangle.points[0].x)
                points[0].Y = intround(triangle.points[0].y)
                points[1].X = intround(triangle.points[1].x)
                points[1].Y = intround(triangle.points[1].y)
                points[2].X = intround(triangle.points[2].x)
                points[2].Y = intround(triangle.points[2].y)
                if u3dsupport.pointInTriangle(point, points):
                    result = triangle.plantPartID
                    return result
        result = closestPartID
        return result
    
    def allocateTriangle(self):
        result = u3dsupport.KfTriangle()
        self.triangles.append(result)
        return result
    
    def sortTrianglesConsideringTdos(self):
        if not self.triangles:
            return
        if self.sortTdosAsOneItem:
            for triangle in self.triangles:
                if not triangle.isLine:
                    oldValue = triangle.zForSorting
                    if triangle.tdo != None:
                        try:
                            # the 0.01 is to keep the polygons inside the tdo in order but still (usually) separate from other tdos
                            # could produce errors if the z value is really small already
                            triangle.zForSorting = triangle.tdo.zForSorting + triangle.zForSorting * 0.01
                        except:
                            triangle.zForSorting = oldValue
                    # have to clear out tdo pointer afterward because triangles are reused
                    triangle.tdo = None
        self.sortTriangles(0, len(self.triangles)-1)
    
    def sortTriangles(self, left, right):
        if right > left:
            z = self.triangles[right].zForSorting
            i = left - 1
            j = right
            while True:
                while 1:
                    i = i + 1
                    if not self.triangles[i].zForSorting < z:
                        break
                while 1:
                    j = j - 1
                    if (j < i) or (not self.triangles[j].zForSorting > z):
                        break
                if i >= j:
                    break
                self.triangles[i], self.triangles[j] = self.triangles[j], self.triangles[i]
            self.triangles[i], self.triangles[right] = self.triangles[right], self.triangles[i]
            self.sortTriangles(left, j)
            self.sortTriangles(i + 1, right)
    
    # ----------------------------------------------------------------------------- KfDrawingSurface drawing 
    def trianglesDraw(self):
        for triangle in self.triangles:
            self.basicDrawTriangle(triangle)
    
    def drawLineFromTo(self, startPoint, endPoint):
        result = None
        if self.recording:
            #store triangle
            triangle = self.allocateTriangle()
            triangle.points[0] = startPoint
            triangle.points[1] = endPoint
            triangle.isLine = True
            #need to handle drawing surface colors
            triangle.foreColor = self.foreColor
            triangle.backColor = self.backColor
            triangle.lineColor = self.lineColor
            triangle.lineWidth = self.lineWidth
            triangle.computeZ()
            result = triangle
        else:
            self.basicDrawLineFromTo(startPoint, endPoint)
        return result
    
    def basicDrawLineFromTo(self, startPoint, endPoint):
        if self.drawingContext == None:
            #assuming pen color and width is saved and restored elsewhere
            return
        #do nothing if rounding fails
        try:
            x1 = intround(startPoint.x)
            y1 = intround(startPoint.y)
            x2 = intround(endPoint.x)
            y2 = intround(endPoint.y)
            width = intround(self.lineWidth)
            self.drawingContextDrawSolidLine(x1, y1, x2, y2, width, self.lineColor)
        except:
            # pdf raise for now
            raise
            pass
    
    def draw3DFace(self, point1, point2, point3, point4):
        # subclasses must override
        raise GeneralException.create("Problem: draw3DFace method not supported in base class; in method KfDrawingSurface.draw3DFace.")
    
    #terminology is different: drawing surface frontColor is pen line color 
    # and drawing surface backColor (which should be back-facing triangle) is fill color
    #drawing surface draws the last triangle allocated - and deallocates it if needed
    def drawLastTriangle(self):
        #last allocated triangle
        triangle = self.triangles[-1]
        triangle.isLine = False
        triangle.updateGeometry()
        #PDF FIX - maybe should scale this?
        triangle.lineWidth = self.lineWidth
        triangle.lineColor = self.lineColor
        triangle.foreColor = self.foreColor
        triangle.backColor = self.backColor
        if self.recording:
            pass
            #"store triangle"
            #do nothing as already stored
        else:
            self.basicDrawTriangle(triangle)
            #deallocate it
            del self.triangles[-1]
    
    #pdf - draw the triangle object - assume it is reasonably flat"
    #{can only draw triangles for now
    #needs fixed size for array - would need to allocate for variable sizes
    #or could use direct windows triangle call and larger array but pass number of points
    #terminology is different: drawing surface frontColor is pen line color
    #and drawing surface backColor (which should be back-facing triangle) is fill color"
    #restricting range of coordingates to 32767 to prevent possible out of range errors when drawing
    def basicDrawTriangle(self, triangle):
        if self.drawingContext == None:
            return
        
        if (triangle.isLine):
            #do nothing if round fails or assign to integer fails
            try:
                #draw line
                startPoint = triangle.points[0]
                endPoint = triangle.points[1]
                x1 = intround(startPoint.x)
                y1 = intround(startPoint.y)
                x2 = intround(endPoint.x)
                y2 = intround(endPoint.y)
                width = intround(triangle.lineWidth)
                self.drawingContextDrawSolidLine(x1, y1, x2, y2, width, triangle.lineColor)
            except:
                # PDF FIX raise for now
                raise
                pass
            return
        
        pointArray = []
        for i in range(4):
            pointArray.append(delphi_compatability.TPoint()) # not sure what type of point it shoudl be
        
        # build point array
        for i in range(0, 4):
            if i < 3:
                # make last point same as first for drawing lines around whole polygon
                aPoint = triangle.points[i]
            else:
                aPoint = triangle.points[0]
            try:
                pointArray[i].X = intround(aPoint.x)
                pointArray[i].Y = intround(aPoint.y)
            except:
                #maybe could test sign to put in negative big numbers? watch out for NaNs!
                pointArray[i].X = 32767
                pointArray[i].Y = 32767
        
        interiorColor = triangle.visibleSurfaceColor()
        
        if self.drawingLines:
            if self.fillingTriangles:
                edgeColor = triangle.drawLinesColor(self.lineContrastIndex)
            else:
                edgeColor = triangle.visibleSurfaceColor()
        else:
            edgeColor = triangle.visibleSurfaceColor()
            
        if self.circlingPoints:
            vertextPointColor = self.circleColor
        else:
            vertextPointColor = None
        
        self.drawingContextDrawTriangle(pointArray, interiorColor, edgeColor, 1, self.fillingTriangles, vertextPointColor)

    # ----------------------------------------------------------------------------- KfDrawingSurface recording 
    def recordingStart(self):
        self.clearTriangles()
        self.recording = True
    
    def recordingStop(self):
        self.recording = False
    
    def recordingDraw(self):
        self.sortTrianglesConsideringTdos()
        self.trianglesDraw()
        
    ################# SUBCLASSES MUST IMPLEMENT
    
    #def drawingContextUseSolidBrush(self):
    #    #self.pane.Brush.Style = delphi_compatability.TFPBrushStyle.bsSolid
    #    raise "Subclass must implement"
    
    #def drawingContextUseClearBrush(self):
    #    #self.pane.Brush.Style = delphi_compatability.TFPBrushStyle.bsClear
    #    raise "Subclass must implement"
    
    def drawingContextDrawSolidLine(self, x1, y1, x2, y2, width, color):
        raise "Subclass must implement"
        # Ideas:
        #self.pane.Pen.Width = intround(self.lineWidth)
        #self.pane.Pen.Color = self.lineColor
        #self.pane.Pen.Style = delphi_compatability.TFPPenStyle.psSolid
        #self.pane.MoveTo(intround(startPoint.x), intround(startPoint.y))
        #self.pane.LineTo(intround(endPoint.x), intround(endPoint.y))
    
    def drawingContextDrawTriangle(self, points, interiorColor, edgeColor, edgeWidth, drawFilled, vertextPointColor=None):
        raise "Subclass must implement"
        # IDEAS:
        #self.pane.Pen.Color = triangle.visibleSurfaceColor()
        #self.pane.Pen.Style = delphi_compatability.TFPPenStyle.psSolid
        #self.pane.Brush.Color = triangle.visibleSurfaceColor()    
        #self.pane.Pen.Width = 1
        
        #if self.fillingTriangles:
        #    self.pane.Polygon(points)
        #else:
        #    self.pane.Polyline(points)
            
        #if self.circlingPoints:
        #    self.pane.Pen.Color = self.circleColor
        #    self.pane.Brush.Color = self.circleColor
        #    for i in range(0, 3):
        #        self.pane.Ellipse(points[i].X - self.circlePointRadius, pointAs[i].Y - self.circlePointRadius, points[i].X + self.circlePointRadius, points[i].Y + self.circlePointRadius)

File: converted_source/used/Udrawingsurface.pas
unit Udrawingsurface;

interface

uses Windows, Graphics, SysUtils, 
  ucollect, utdo, u3dsupport;

const
  kInitialDrawingSurfaceTriangles = 1000;

type

	KfDrawingSurface = class(TObject)
    public
    pane: TCanvas;
    lineContrastIndex: smallint;
    drawingLines, fillingTriangles, circlingPoints: boolean;
    foreColor, backColor, lineColor, circleColor: TColor;
    numTrianglesUsed: longint;
    triangles: TListCollection;
    recording: boolean;
    lineWidth: single;
    circlePointRadius: smallint;
    sortTdosAsOneItem: boolean;
    constructor create; virtual;
    destructor destroy; override;
    procedure draw3DFace(point1, point2, point3, point4: KfPoint3D); virtual;
		procedure basicDrawLineFromTo(const startPoint: KfPoint3D; const endPoint: KfPoint3D); virtual;
		procedure basicDrawTriangle(triangle: KfTriangle); virtual;
    procedure clearTriangles;
		function allocateTriangle: KfTriangle;
		procedure sortTrianglesConsideringTdos;
		procedure sortTriangles(left: integer; right: integer);
		procedure recordingStart;
		procedure recordingStop;
		procedure trianglesDraw;
 		procedure recordingDraw;
    function drawLineFromTo(const startPoint: KfPoint3D; const endPoint: KfPoint3D): KfTriangle;
		procedure drawLastTriangle;
		procedure initialize;
    function plantPartIDForPoint(point: TPoint): longint;
  	end;

implementation

{ ----------------------------------------------------------------------------- *KfDrawingSurface creating/destroying }
constructor KfDrawingSurface.create;
  var i: longint;
	begin
  inherited create;
  triangles := TListCollection.create;
  numTrianglesUsed := 0;
  fillingTriangles := true;
  self.initialize;
  { start with a bunch of triangles }
  for i := 0 to kInitialDrawingSurfaceTriangles do
    triangles.add(KfTriangle.create);
  end;

destructor KfDrawingSurface.destroy;
	begin
  triangles.free;
  inherited destroy;
  end;

procedure KfDrawingSurface.initialize;
  begin
  {backColor := clLtGray;
  foreColor := clLtGray;
  lineColor := clLtGray;}
  lineWidth := 1.0;
  recording := false;
  clearTriangles;
  if self.pane <> nil then
    begin
    if fillingTriangles then
      self.pane.brush.style := bsSolid
    else
      self.pane.brush.style := bsClear;
    end;
  end;

{ ----------------------------------------------------------------------------- KfDrawingSurface managing triangles }
procedure KfDrawingSurface.clearTriangles;
  var i: longint;
  begin
  for i := 0 to numTrianglesUsed - 1 do
    KfTriangle(triangles.items[i]).tdo := nil;
  numTrianglesUsed := 0;
  end;

function KfDrawingSurface.plantPartIDForPoint(point: TPoint): longint;
  var
    i: longint;
    triangle: KfTriangle;
    points: array[0..2] of TPoint;
    widenedPoints: array[0..2] of TPoint;
    width: integer;
    thisDistance, closestDistanceSoFar: single;
    closestPartID: longint;
    centerX, centerY, x, y: single;
  begin
  x := point.x;
  y := point.y;
  result := -1;
  closestDistanceSoFar := 0;
  closestPartID := -1;
  for i := 0 to triangles.count - 1 do
    begin
    triangle := KfTriangle(triangles.items[i]);
    if triangle.isLine then
      begin
      centerX := (triangle.points[0].x + triangle.points[1].x) / 2.0;
      centerY := (triangle.points[0].y + triangle.points[1].y) / 2.0;
      thisDistance := (centerX - x) * (centerX - x) + (centerY - y) * (centerY - y);
      if (closestPartID = -1) or (thisDistance < closestDistanceSoFar) then
        begin
        closestDistanceSoFar := thisDistance;
        closestPartID := triangle.plantPartID;
        end;
      end
    else
      begin
      points[0].x := round(triangle.points[0].x); points[0].y := round(triangle.points[0].y);
      points[1].x := round(triangle.points[1].x); points[1].y := round(triangle.points[1].y);
      points[2].x := round(triangle.points[2].x); points[2].y := round(triangle.points[2].y);
      if pointInTriangle(point, points) then
        begin
        result := triangle.plantPartID;
        exit;
        end;
      end;
     end;
   result := closestPartID;
  end;

function KfDrawingSurface.allocateTriangle: KfTriangle;
  begin
  result := nil;
  if numTrianglesUsed < triangles.count then
    result := triangles.items[numTrianglesUsed]
  else
    begin
    result := KfTriangle.create;
    triangles.add(result);
    end;
  inc(numTrianglesUsed);
  end;

procedure KfDrawingSurface.sortTrianglesConsideringTdos;
  var
    i: longint;
    triangle: KfTriangle;
    oldValue: single;
	begin
  if numTrianglesUsed <= 1 then exit;
  if sortTdosAsOneItem then
    for i := 0 to numTrianglesUsed - 1 do
      begin
      triangle := KfTriangle(triangles.items[i]);
      if not triangle.isLine then
        begin
        oldValue := triangle.zForSorting;
        // the 0.01 is to keep the polygons inside the tdo in order but still (usually) separate from other tdos
        // could produce errors if the z value is really small already
        if triangle.tdo <> nil then
          try
            triangle.zForSorting := triangle.tdo.zForSorting + triangle.zForSorting * 0.01;
          except
            triangle.zForSorting := oldValue;
          end;
        // have to clear out tdo pointer afterward because triangles are reused
        triangle.tdo := nil;
        end;
      end;
  self.sortTriangles(0, numTrianglesUsed - 1);
  end;

procedure KfDrawingSurface.sortTriangles(left: integer; right: integer);
  var
    i, j: integer;
    z: single;
  begin
  if right > left then
  	begin
    z := KfTriangle(triangles.items[right]).zForSorting;
    i := left - 1;
    j := right;
    while true do
      begin
      repeat
        i := i + 1;
        until not (KfTriangle(triangles.items[i]).zForSorting < z);
      repeat
        j := j - 1;
        until (j < i) or (not (KfTriangle(triangles.items[j]).zForSorting > z));
      if i >= j then break;
      triangles.exchange(i,j);
      end;
    triangles.exchange(i, right);
    sortTriangles(left, j);
    sortTriangles(i + 1, right);
    end;
	end;

{ ----------------------------------------------------------------------------- KfDrawingSurface drawing }
procedure KfDrawingSurface.trianglesDraw;
	var i: integer;
	begin
  if numTrianglesUsed > 0 then
  	for i := 0 to numTrianglesUsed - 1 do
    	self.basicDrawTriangle(triangles.items[i]);
  end;

function KfDrawingSurface.drawLineFromTo(const startPoint: KfPoint3D; const endPoint: KfPoint3D): KfTriangle;
  var triangle: KfTriangle;
  begin
  result := nil;
  if self.recording then
  	begin
    {store triangle}
    triangle := self.allocateTriangle;
    triangle.points[0] := startPoint;
    triangle.points[1] := endPoint;
    triangle.isLine := true;
    {need to handle drawing surface colors}
    triangle.foreColor := foreColor;
    triangle.backColor := backColor;
    triangle.lineColor := lineColor;
    triangle.lineWidth := lineWidth;
    triangle.computeZ;  
    result := triangle;
    end
  else
    self.basicDrawLineFromTo(startPoint, endPoint);
  end;

procedure KfDrawingSurface.basicDrawLineFromTo(const startPoint: KfPoint3D; const endPoint: KfPoint3D);
  begin
  {assuming pen color and width is saved and restored elsewhere}
  if self.pane = nil then exit;
  try
    self.pane.pen.width := round(self.lineWidth) mod 1000;
    self.pane.pen.color := self.lineColor;
    self.pane.pen.style := psSolid;
    self.pane.moveTo(round(startPoint.x) mod 32767, round(startPoint.y) mod 32767);
    self.pane.lineTo(round(endPoint.x) mod 32767, round(endPoint.y) mod 32767);
  except
    {do nothing if rounding fails}
  end;
  end;

procedure KfDrawingSurface.draw3DFace(point1, point2, point3, point4: KfPoint3D);
  begin
  // subclasses must override
  raise exception.create('Problem: draw3DFace method not supported in base class; in method KfDrawingSurface.draw3DFace.');
  end;

{terminology is different: drawing surface frontColor is pen line color }
{ and drawing surface backColor (which should be back-facing triangle) is fill color}
{drawing surface draws the last triangle allocated - and deallocates it if needed}
procedure KfDrawingSurface.drawLastTriangle;
  var triangle: KfTriangle;
  begin
  triangle := triangles.items[numTrianglesUsed - 1]; {last allocated triangle}
  triangle.isLine := false;
  triangle.updateGeometry;
  triangle.lineWidth := lineWidth; {PDF FIX - maybe should scale this?}
  triangle.lineColor := lineColor;
  triangle.foreColor := foreColor;
  triangle.backColor := backColor;
  if self.recording then
  	begin
    {"store triangle"}
    {do nothing as already stored}
    end
  else
  	begin
    self.basicDrawTriangle(triangle);
    {deallocate it}
    dec(numTrianglesUsed);
    end;
  end;

{pdf - draw the triangle object - assume it is reasonably flat"
{can only draw triangles for now}
{needs fixed size for array - would need to allocate for variable sizes}
{or could use direct windows triangle call and larger array but pass number of points}
{terminology is different: drawing surface frontColor is pen line color}
{and drawing surface backColor (which should be back-facing triangle) is fill color"}
{restricting range of coordingates to 32767 to prevent possible out of range errors when drawing}
procedure KfDrawingSurface.basicDrawTriangle(triangle: KfTriangle);
  var
    startPoint: KfPoint3D;
    endPoint: KfPoint3D;
    aPoint: KfPoint3D;
    pointArray: array[0..3] of TPoint;
    i: integer;
  begin
  if self.pane = nil then exit;
  if (triangle.isLine) then
    begin
    {draw line}
    try
      self.pane.pen.width := round(triangle.lineWidth);
      self.pane.pen.color := triangle.lineColor;
      self.pane.pen.style := psSolid;
      startPoint := triangle.points[0];
      endPoint := triangle.points[1];
      self.pane.moveTo(round(startPoint.x) mod 32767, round(startPoint.y) mod 32767);
      self.pane.lineTo(round(endPoint.x) mod 32767, round(endPoint.y) mod 32767);
    except
      {do nothing if round fails or assign to integer fails}
    end;
    exit;
    end;
  for i := 0 to 3 do
    begin
    // make last point same as first for drawing lines around whole polygon
    if i < 3 then
      aPoint := triangle.points[i]
    else
      aPoint := triangle.points[0];
    try
      pointArray[i].x := round(aPoint.x) mod 32767;
      pointArray[i].y := round(aPoint.y) mod 32767;
    except
      {maybe could test sign to put in negative big numbers? watch out for NaNs!}
      pointArray[i].x := 32767;
      pointArray[i].y := 32767;
    end;
    end;
  self.pane.pen.style := psSolid;
  self.pane.brush.color := triangle.visibleSurfaceColor;
  if self.drawingLines then
    begin
    if fillingTriangles then
      self.pane.pen.color := triangle.drawLinesColor(lineContrastIndex)
    else
      self.pane.pen.color := triangle.visibleSurfaceColor;
    end
  else
    self.pane.pen.color := triangle.visibleSurfaceColor;
  self.pane.pen.width := 1;
  if fillingTriangles then
    self.pane.polygon(pointArray)
  else
    self.pane.polyLine(pointArray);   
  if self.circlingPoints then
    begin
    self.pane.pen.color := circleColor;
    self.pane.brush.color := circleColor;
    for i := 0 to 2 do
      self.pane.ellipse(pointArray[i].x - circlePointRadius, pointArray[i].y - circlePointRadius,
          pointArray[i].x + circlePointRadius, pointArray[i].y + circlePointRadius);
    end;
	end;

{ ----------------------------------------------------------------------------- KfDrawingSurface recording }
procedure KfDrawingSurface.recordingStart;
	begin
  self.clearTriangles;
  recording := true;
  end;

procedure KfDrawingSurface.recordingStop;
	begin
  recording := false;
  end;

procedure KfDrawingSurface.recordingDraw;
	begin
  self.sortTrianglesConsideringTdos;
  self.trianglesDraw;
  end;

end.
 